[ETHEREUM-CONTRACTS] GDA

0xdavinchee opened the pull request:
## Rationale
This PR includes the implementation of the GDA, a new agreement to be added to the protocol.

## Summary of Changes
This PR is fairly large so instead of trying to review everything in a single sitting, it is more effective to split it up and review based on "themes".
1. Core | Agreement Contract - GDA & SuperfluidPool: can be broken down into first looking at the APIs and basics then questioning security assumptions
2. Peripherals | Proxy and Upgradability
3. Peripherals | Flow NFT changes & Pool NFTs
4. Deployments | deploy-framework - bootstrap & upgrade paths
5. Deployments | print info and contract verification
6. Deployments | dev scripts
7. Tests | Property tests
8. Tests | Integration tests

hellwolf commented on packages/automation-contracts/scheduler/foundry.toml:
```diff
@@ -3,6 +3,7 @@ root = '../../../'
 libs = ['lib']
 src = 'packages/automation-contracts/scheduler'
 remappings = [
+    '@superfluid-finance/solidity-semantic-money/src/=packages/solidity-semantic-money/src/',
```
looks like this could be changes in a different PR?

hellwolf commented on .github/workflows/ci.feature.yml:
```diff
@@ -129,7 +129,7 @@ jobs:
           lcov -e lcov.info 'packages/ethereum-contracts/contracts/*' -o lcov.info
           # remove mocks, base super app, test and deployer contracts (see .solcover.js)
-          lcov -r lcov.info 'packages/ethereum-contracts/contracts/mocks/*' 'packages/ethereum-contracts/contracts/apps/*Base*' 'packages/ethereum-contracts/contracts/utils/*Test*' 'packages/ethereum-contracts/contracts/utils/*Deployer*' -o lcov.info
+          lcov -r lcov.info 'packages/ethereum-contracts/contracts/mocks/*' 'packages/ethereum-contracts/contracts/apps/*Base*' 'packages/ethereum-contracts/contracts/utils/*Test*' 'packages/ethereum-contracts/contracts/utils/*Deploy*' -o lcov.info
```
maybe also a separate improvement PR?

0xdavinchee committed:
Merge branch 'dev' into gda

hellwolf commented:
Changing title, this is not a drill :) : [ETHEREUM-CONTRACTS] GDA ~Prototype~

0xdavinchee committed:
SuperfluidUpgradeableBeaconTest test contract

0xdavinchee committed:
adjustment fr

0xdavinchee committed:
Update GeneralDistributionAgreement.t.sol

0xdavinchee committed:
more testing cleanup

0xdavinchee committed:
we'll do it LIVE!

- deployments uncommented
- test cleanups

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/agreements/IGeneralDistributionAgreementV1.sol:
```diff
@@ -0,0 +1,156 @@
+    error GDA_ONLY_SUPER_TOKEN_POOL();              // 0x90028c37
+
+    // Events
+    event InstantDistributionUpdated(
```
what about address "operator"?

0xdavinchee replied:
added

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/agreements/IGeneralDistributionAgreementV1.sol:
```diff
@@ -0,0 +1,156 @@
+    event FlowDistributionUpdated(
+        ISuperfluidToken indexed token,
+        ISuperfluidPool indexed pool,
+        address operator,
```
an extra code describing why operator had permission?

- due to liquidation
- due to ACL

0xdavinchee replied:
resolved

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/agreements/IGeneralDistributionAgreementV1.sol:
```diff
@@ -0,0 +1,156 @@
+    event PoolCreated(ISuperfluidToken indexed token, address indexed admin, ISuperfluidPool pool);
+
+    event PoolConnectionUpdated(
+        ISuperfluidToken indexed token, address indexed account, ISuperfluidPool indexed pool, bool connected
```
order of event params: token, pool, ...

0xdavinchee replied:
resolved

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/agreements/IGeneralDistributionAgreementV1.sol:
```diff
@@ -0,0 +1,156 @@
+        ISuperfluidToken indexed token, address indexed account, ISuperfluidPool indexed pool, bool connected
+    );
+
+    event UniversalIndexUpdated(
```
probably too internal detail

0xdavinchee replied:
resolved

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/agreements/IGeneralDistributionAgreementV1.sol:
```diff
@@ -0,0 +1,156 @@
+    /// @param to The pool address
+    /// @param requestedFlowRate The requested flow rate
+    /// @return actualFlowRate
+    function getFlowDistributionActualFlowRate(
```
rename to estimateFlowDistributionActualFlowRate

Note: it is indeed an estimation, because total units could be updated in between (although only the admin can update it).

hellwolf replied:
Also fix the comments

0xdavinchee replied:
addressed

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/agreements/IGeneralDistributionAgreementV1.sol:
```diff
@@ -0,0 +1,156 @@
+        address from,
+        ISuperfluidPool to,
+        int96 requestedFlowRate
+    ) external view virtual returns (int96 actualFlowRate);
```
maybe return also the actualTotalDistributionFlowRate.

0xdavinchee replied:
addressed

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/agreements/IGeneralDistributionAgreementV1.sol:
```diff
@@ -0,0 +1,156 @@
+        uint256 distributedAtTimestamp,
+        int96 oldFlowRate,
+        int96 newDistributorToPoolFlowRate,
+        int96 newTotalDistributionFlowRate
```
adjustment flow related info can also be added here:

- adjustmentFlowReceiver,
- adjustmentFlowRate

0xdavinchee replied:
addressed

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/agreements/IGeneralDistributionAgreementV1.sol:
```diff
@@ -0,0 +1,156 @@
+    /// @param token The token address
+    /// @param account The account address
+    /// @return true if `account` is a pool
+    function isPool(ISuperfluidToken token, address account) external view virtual returns (bool);
```
missing isMemberConnected

0xdavinchee replied:
added

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidPool.sol:
```diff
@@ -0,0 +1,91 @@
+    error SUPERFLUID_POOL_NOT_GDA();        // 0xfcbe3f9e
+
+    // Events
+    event MemberUpdated(address indexed member, uint128 units, uint256 updatedAt);
```
updatedAt seems redundant.

0xdavinchee replied:
addressed

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidPool.sol:
```diff
@@ -0,0 +1,91 @@
+
+    // Events
+    event MemberUpdated(address indexed member, uint128 units, uint256 updatedAt);
+    event PoolIndexUpdated(uint128 totalUnits, uint32 wpSettledAt, int256 wpSettledValue, int96 wpFlowRate);
```
internal information, can be hidden.

0xdavinchee replied:
addressed

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidPool.sol:
```diff
@@ -0,0 +1,91 @@
+    // Events
+    event MemberUpdated(address indexed member, uint128 units, uint256 updatedAt);
+    event PoolIndexUpdated(uint128 totalUnits, uint32 wpSettledAt, int256 wpSettledValue, int96 wpFlowRate);
+    event DistributionClaimed(address indexed member, int256 claimableAmount, int256 totalClaimed, uint256 timestamp);
```
ditto re timestamp

0xdavinchee replied:
resolved

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidPool.sol:
```diff
@@ -0,0 +1,91 @@
+    event PoolIndexUpdated(uint128 totalUnits, uint32 wpSettledAt, int256 wpSettledValue, int96 wpFlowRate);
+    event DistributionClaimed(address indexed member, int256 claimableAmount, int256 totalClaimed, uint256 timestamp);
+
+    /// @notice The pool admin
```
some additional comments about what admin can have additionally: units updating privilege, and adjustment flow rate recipient.

0xdavinchee replied:
addressed

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidPool.sol:
```diff
@@ -0,0 +1,91 @@
+    function getTotalUnits() external view returns (uint128);
+
+    /// @notice The total number of units of disconnected members
+    function getDisconnectedUnits() external view returns (uint128);
```
getConnectedUnits()

hellwolf replied:
alternative: getTotalDisconnectedUnits getTotalConnectedUnits

0xdavinchee replied:
addressed

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidPool.sol:
```diff
@@ -0,0 +1,91 @@
+    function getUnits(address memberAddress) external view returns (uint128);
+
+    /// @notice The flow rate of the connected members
+    function getConnectedFlowRate() external view returns (int96);
```
alternative for consistency: getTotalDisconnectedFlowRate & getTotalConnectedFlowRate

0xdavinchee replied:
addressed

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidPool.sol:
```diff
@@ -0,0 +1,91 @@
+
+    /// @notice The balance of all the disconnected members at `time`
+    /// @param time The time to query
+    function getDisconnectedBalance(uint32 time) external view returns (int256 balance);
```
alternative: getTotalDisconnectedBalance

hellwolf replied:
1) getTotalDisconnectedBalance = forall m. isDisconnected m => sum m.getClaimable()

2) forall m. isConnected m ==> m.getClaimable() == 0

=>

if 2) is guaranteed, then instead, we should just have the `getTotalClaimable` API instead:

3) getTotalClaimable = forall m. sum m.getClaimable()

=>

additionally,

4) in all cases

a = m.balance;
b = m.getClaimable()
m.claimAll();
c = m.balance;
assert(c ==  a + b);


-----

Independently,

5) in all cases <------

when t.isConnected(p, m):
  a = m.balance
  t.disconnectPool(p, m)
  b = m.balance
  assert(a == b)

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidPool.sol:
```diff
@@ -0,0 +1,91 @@
+    function operatorConnectMember(address memberAddr, bool doConnect, uint32 time) external returns (bool);
+}
+
+interface ISuperfluidPoolAdmin {
```
This should be ISuperfluidPoolOperator

hellwolf replied:
~~We can remove this one from public interface, and keep it part of the GDA implementation instead; so only Pool implementation needs to use it directly (Pool implementation already directly uses through GDA)~~

Move some of these ISuperfluidPoolAdmin to IGDA instead? And some can be part of the GDA directly.

0xdavinchee replied:
addressed

hellwolf commented on packages/ethereum-contracts/contracts/interfaces/superfluid/ISuperfluidPool.sol:
```diff
@@ -0,0 +1,91 @@
+
+    /// @notice Does a claim for `memberAddr` at `time` and handles internal accounting logic
+    /// @dev This is only callable by the GDA and is not to be used by the end user
+    function operatorConnectMember(address memberAddr, bool doConnect, uint32 time) external returns (bool);
```
move this operatorConnectMember to the implementation of the SuperfluidPool.

0xdavinchee replied:
addressed

0xdavinchee committed:
formatting

hellwolf commented on packages/ethereum-contracts/contracts/mocks/GeneralDistributionAgreementV1Mock.sol:
```diff
@@ -0,0 +1,130 @@
+/// @title GeneralDistributionAgreementV1Mock
+/// @author Superfluid
+/// @notice A GDAv1 Mock contract which exposes internal functions for testing purposes
+contract GeneralDistributionAgreementV1Mock is GeneralDistributionAgreementV1 {
```
For future test plan fine tuning, Mock here should be more accurately called "Harness".

0xdavinchee replied:
addressed

hellwolf commented on packages/ethereum-contracts/contracts/superfluid/SuperfluidPool.sol:
```diff
@@ -0,0 +1,293 @@
+    using SafeCast for uint256;
+    using SafeCast for int256;
+
+    struct PoolIndexData {
```
missing storage layout compatibility validation code for upgradability guarantee check.

hellwolf commented on packages/ethereum-contracts/contracts/superfluid/SuperfluidPool.sol:
```diff
@@ -0,0 +1,293 @@
+        int256 claimedValue;
+    }
+
+    GeneralDistributionAgreementV1 internal immutable _gda;
```
make it public: GDA

0xdavinchee replied:
resolved

hellwolf commented on packages/ethereum-contracts/contracts/superfluid/SuperfluidPool.sol:
```diff
@@ -0,0 +1,293 @@
+    }
+
+    function proxiableUUID() public pure override returns (bytes32) {
+        return keccak256("org.superfluid-finance.contracts.superfluid.SuperfluidPool.implementation");
```
-> SuperfluidPool.v1 to be consistent with GDA v1

0xdavinchee replied:
what is currently there is more consistent with other proxiableUUID function strings, e.g. `org.superfluid-finance.contracts.SuperfluidGovernanceII.implementation`

hellwolf commented on packages/ethereum-contracts/contracts/superfluid/SuperfluidPool.sol:
```diff
@@ -0,0 +1,293 @@
+
+    /// @inheritdoc ISuperfluidPool
+    function getDisconnectedUnits() external view override returns (uint128) {
+        return _disconnectedMembers.ownedUnits;
```
hear hear

0xdavinchee replied:
? I guess this is changed to `getTotalDisconnectedUnits`

hellwolf commented on packages/ethereum-contracts/contracts/superfluid/SuperfluidPool.sol:
```diff
@@ -0,0 +1,293 @@
+        });
+    }
+
+    function convertPoolIndexDataToPDPoolIndex(PoolIndexData memory data)
```
make it internal

hellwolf commented on packages/ethereum-contracts/contracts/superfluid/SuperfluidPool.sol:
```diff
@@ -0,0 +1,293 @@
+        return true;
+    }
+
+    function operatorSetIndex(PDPoolIndex calldata index) external onlyGDA returns (bool) {
```
:+1: onlyGDA

hellwolf commented on packages/ethereum-contracts/contracts/superfluid/SuperfluidPool.sol:
```diff
@@ -0,0 +1,293 @@
+
+    /// @inheritdoc ISuperfluidPool
+    function updateMember(address memberAddr, uint128 newUnits) external returns (bool) {
+        if (admin != msg.sender) revert SUPERFLUID_POOL_NOT_POOL_ADMIN();
```
:+1: only admin

hellwolf commented on packages/ethereum-contracts/contracts/superfluid/SuperfluidPool.sol:
```diff
@@ -0,0 +1,293 @@
+    /// @inheritdoc ISuperfluidPool
+    function claimAll(address memberAddr) public returns (bool) {
+        bool isConnected = _gda.isMemberConnected(superToken, address(this), memberAddr);
+        uint32 time = uint32(block.timestamp);
```
please add a TODO, GDA.getHost().getTimestamp() should be used in principle

0xdavinchee replied:
resolved

hellwolf commented on packages/ethereum-contracts/contracts/superfluid/SuperfluidPool.sol:
```diff
@@ -0,0 +1,293 @@
+    }
+
+    // WARNING for operators: it is undefined behavior if member is already connected or disconnected
+    function operatorConnectMember(address memberAddr, bool doConnect, uint32 time) external onlyGDA returns (bool) {
```
+1 onlyGDA

hellwolf commented on packages/ethereum-contracts/contracts/upgradability/BeaconProxiable.sol:
```diff
@@ -0,0 +1,17 @@
+
+abstract contract BeaconProxiable is Initializable {
+
+    // allows to mark logic contracts as initialized in order to reduce the attack surface
```
> reduce the attack surface

->

-> +, also to simplify the life for bug bounty white hacker by asking them to look else where. Because we understand that there should be little to no attack surface without initializer.

hellwolf replied:
Same to the UUPSProxiable

0xdavinchee replied:
addressed

hellwolf commented on packages/ethereum-contracts/contracts/upgradability/SuperfluidUpgradeableBeacon.sol:
```diff
@@ -0,0 +1,31 @@
+} from "@openzeppelin/contracts/proxy/beacon/UpgradeableBeacon.sol";
+import { BeaconProxiable } from "./BeaconProxiable.sol";
+
+contract SuperfluidUpgradeableBeacon is UpgradeableBeacon {
```
Note for the next review:

Consider to merge SuperfluidUpgradeableBeacon & BeaconProxiable into BeaconProxiable, similar to how UUPSProxy is so that more code can be in the upgradable area.

hellwolf reviewed (commented):
GDA Review Session #1

0xdavinchee committed:
wip cleanups + test removals

- cleanup interface
- add estimateDistributionActualAmount
- remove a bunch of the gda demo specific tests

0xdavinchee committed:
address some review comments

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
review comments cont.

0xdavinchee committed:
fix build

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
should fix build + allow deploy

0xdavinchee committed:
fix deployments

- renames of supertokenpool => superfluidpool
- fix deploy-framework

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
GDA wip

- clean up functions
- add isPatricianPeriod view functions
- fix distributeFlow shifting order: adjust buffer happens after flow distribution
- add basic liquidation test
- increase coverage
- make _assertGlobalInvariants virtual

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
flow nft cleanup/fixes

- add metadata to ethereum-contracts
- FlowNFT's do not take baseURI in constructor, it is a constant variable now, no longer in storage
- SuperTokenFactory fixed to check existing canonical addresses against new contract canonical logic addresses
- add validate nft deployments addresses script

0xdavinchee committed:
workflow cleanups

- add `permissions: write-all` to broken workflows
- fix deploy-framework.js up
- fix scripts => dev-scripts

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Merge branch 'flow-nft-deployments-fixes' into gda

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
fix up framework deployment

- add more validation to script
- proxiableUUID fixed for SuperfluidPool
- deploy-framework fix

0xdavinchee committed:
wip

- add trusted forwarder test
- fix info-print-contract-addresses

0xdavinchee committed:
cleanup

- optimization in distributeFlow
- added more test cases

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Pool NFTs WIP

- IPoolAdminNFT, IPoolMemberNFT, IPoolNFTBase added
- ISuperToken interface modified to include pool nft's
- SuperToken and SuperTokenFactoryMock modified given updates to constructor
- ConstantOutflowNFT: CFA or GDA
- PoolAdminNFT, PoolMemberNFT and PoolNFTBase initial implementation
- SuperToken constructor takes flow NFT proxies
- SuperTokenFactory constructor takes flow NFT logic contracts
  - update code path also handles nft logic upgrades
- SF deployment steps modified to deploy flow NFTs
- deployment scripts modified: deploy-framework, info-print-addresses, verify-framework
- other test files modified given new constructor
- Fix up existing TS test suite

0xdavinchee committed:
add erc20 to superfluidpool

0xdavinchee committed:
pool nft integration

- mint/burn pool member nft in superfluidpool: update member units
- mint pool admin nft in gda: create pool

0xdavinchee committed:
gda wip

- inc/dec allowance
- move IERC20 to ISuperfluidPool
- add helpers + basic tests for approve/transfer
- transfer WIP

0xdavinchee committed:
validate + make test more robust

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
gda tests cleanup

0xdavinchee committed:
fix verification

- always use slotsBitmapLibraryAddress deployed for IDA for GDA

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
flow nft on distribute flow

0xdavinchee committed:
testing cleanup

0xdavinchee committed:
fix transfer test too

0xdavinchee committed:
GDAv1 in SuperTokenV1Library

- remove andAssert functions
- use startPrank
- clean up tests

0xdavinchee committed:
full coverage

0xdavinchee committed:
wip

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Update deploy-framework.js

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Update ISuperToken.sol

0xdavinchee committed:
using lib for ISuperToken

0xdavinchee committed:
add tests for lib

- adding tests for library

0xdavinchee committed:
gda wip

- NFTs (Admin and Member) mint functions require passed pool to be registered
- more function reuse in superfluidpool
- add update function in pool member nft and call it in SuperfluidPool (similar to ConstantOutflowNFT.onUpdate)
- use uint64 max flow rate in tests instead of int32

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
fix build

0xdavinchee committed:
function renaming + comments

0xdavinchee committed:
gda harness cleanup

- reduce code size of gdav1harness by combining function
- fix up tests
- add pool param to adjustBuffer
- add superToken to superfluid pool events
- new BufferAdjusted event in GDAv1

0xdavinchee committed:
gda subgraph hemingway bridge

- gda schema added
- getAbi.js cleaned up
- manifest file updated
- mapping skeleton added (not for aggregate yet)

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
gdav1 forwarder wip

0xdavinchee committed:
mapping functions WIP

- schema cleanup
- event entities mapping implemented
- hol entities mapping implemented

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
aggregate entities schema

0xdavinchee committed:
- add docs for new properties
- initialize new properties in getOrInit
- add tests for new properties in matchstick

0xdavinchee committed:
Delete hardhat.json

0xdavinchee committed:
nft's wip

- pool member nft try/catch
- pool admin nft try/catch
- make the to parameter an ISuperfluidPool type and fix expected types in tests and peripherals

0xdavinchee committed:
add transfer units to pool test

0xdavinchee committed:
nft testing wip

- move `NoNFTSuperTokenMock` to `SuperTokenMock.sol`
- pool NFT mocks and upgradability mock contracts created
- PoolNFTBase integration tests includes storage layout tests

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
fix build

0xdavinchee committed:
wip + refactoring tests

- moved common NFT test code to ERC721.t.sol
- removed `Proxy` suffix
- added upgradability tests for PoolNFTs

0xdavinchee committed:
fix build

0xdavinchee committed:
testing wip

- discovered some refactoring possibilities
- added some helper functions
- tested nft integration

0xdavinchee committed:
tests refactor

- refactored flow nft tests (constant inflow, constant outflow)
- added tests for pool nft base

0xdavinchee committed:
pool nft tests wip + cleanup

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
add assertions for nft's in gda helpers

0xdavinchee committed:
add GDA to SDK-Core

add gda and SuperfluidPool code to SDK-Core

0xdavinchee committed:
TSDoc for SuperfluidPool

0xdavinchee committed:
GDA in SDK-Core WIP

- add GDA to SuperfluidLoader
- deploy IDAv1Forwarder and GDAv1Forwarder in deploy-framework.js in local testing
- add gdav1 to Framework initialization
- add tsdoc to GeneralDistributionAgreementV1.ts
- connect provider or signer to view contract calls
- connect signer to write contract calls
- add gda and superfluid pool functions to supertoken
- gda tests wip

0xdavinchee committed:
more tests and coverage WIP

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
wip

- clean up forwarder
- clean up deployer (remove duplicate functions)
- add forwarder to deployer
- add reusable function for deploying forwarders in deploy-framework
- fix interfaces.ts for sdk-core

0xdavinchee committed:
fix build + add tests

0xdavinchee committed:
force artifact creation

0xdavinchee committed:
undo force

0xdavinchee committed:
final SDK-Core tests for GDA

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
make yarn dev use forge by default

0xdavinchee committed:
refactor tests

- move GDA helpers and asserts from GDAv1.t.sol to FoundrySuperfluidTester

0xdavinchee committed:
Update package.json

0xdavinchee committed:
tidying up

0xdavinchee committed:
add total connected flow rate to pool

0xdavinchee committed:
fix distribute tests

- minor naming refactoring
- uncomment _helperDistribute in randomPoolSeqs

0xdavinchee committed:
add batch liquidator + toga to Framework struct

0xdavinchee committed:
does not exist error

0xdavinchee committed:
batch liquidator modified to handle GDA

0xdavinchee committed:
fix code size too large

0xdavinchee committed:
updateMemberUnits name change, _toSemanticMoneyUnit added, set TOGA as default reward address

0xdavinchee committed:
toSemanticMoney

0xdavinchee committed:
merge dev

ngmachado committed:
Merge remote-tracking branch 'origin/dev' into gda

0xdavinchee committed:
merge dev

0xdavinchee committed:
testing wip

0xdavinchee committed:
fix build to unblock

hellwolf commented:
I created the `gda-slither` branch, merges both this branch and the `update_nix_flake_inputs` where the fixes to slither and echidna are.

Here is how to test:

```
$ cd packages/ethereum-contracts
$ # create these symlinks
$ ll lib packages
lrwxrwxrwx 1 hellwolf hellwolf  9 Jul 24 12:58 lib -> ../../lib/
lrwxrwxrwx 1 hellwolf hellwolf 14 Jul 24 12:58 packages -> ../../packages/
$ slither . --compile-force-framework foundry --foundry-out-directory build/foundry/out
```

Report is attached to a different issue instead: https://github.com/superfluid-finance/protocol-monorepo/issues/1545

0xdavinchee committed:
fix build

hellwolf committed:
Update SuperTokenV1Library.sol: make isMemberConnected internal

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
cleanup and wip

- move nft code blocks into functions for easier reasoning in gda and pool
- remove check for if flow exists before deletion
- fix test case where lib deploy not needed
- rename test contract
- helper test functions cleanup + additions

0xdavinchee committed:
gda test cleanup

- remove old comments
- remove _expectedPoolData
- assert flow rate for member is correct

0xdavinchee committed:
add assertions for distribute claimable

0xdavinchee committed:
check actual flow rate

- check distributor flow rate against actual flow rate
- fix estimate flow distribution flow rate logic (adjustment flow case)

0xdavinchee committed:
final helper assertion for distribute flow

0xdavinchee committed:
fix build

0xdavinchee committed:
subgraph progress

0xdavinchee committed:
subgraph wip

- MemberUpdated => MemberUnitsUpdated
- gda aggregate entities mapping wip

0xdavinchee committed:
superfluidpool mapping

- finish aggregate mapping for superfluid pool

0xdavinchee committed:
Update mappingHelpers.ts

0xdavinchee committed:
subgraph tests wip

- a bit unsure why the error is occurring in the subgraph tests...

0xdavinchee committed:
fix subgraph tests

- ordering of event params was wrong

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
clean up tests

0xdavinchee committed:
add gda addresses

0xdavinchee committed:
fix tests

0xdavinchee committed:
fix deployment

0xdavinchee committed:
contracts notes

0xdavinchee committed:
tests cleanup

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
uncommented test

0xdavinchee committed:
more tests + helpers added

0xdavinchee committed:
more helpers and abstractions!

- added some more subgraph tests
- fix le build

hellwolf committed:
Merge branch 'dev' into gda

hellwolf committed:
fix lint:sol errors

hellwolf committed:
fix echidna

0xdavinchee committed:
warnings cleanup

hellwolf committed:
break windows

hellwolf committed:
Merge branch 'dev' into keep-up-with-solhint-recommendations

0xdavinchee committed:
add one more test

0xdavinchee committed:
Merge branch 'dev' into keep-up-with-solhint-recommendations

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
use supertokenv1library

- get gda hot fuzz in

hellwolf committed:
fixing warnings

0xdavinchee committed:
Merge branch 'dev' into keep-up-with-solhint-recommendations

hellwolf committed:
fix semantic money

0xdavinchee committed:
clean up build warnings

hellwolf committed:
do not ignore success flag from low level call

hellwolf committed:
disable deny_warnings for now

0xdavinchee committed:
Merge branch 'keep-up-with-solhint-recommendations' into gda

0xdavinchee committed:
fix up build and lint

0xdavinchee committed:
fix build

0xdavinchee committed:
clean up echidna

0xdavinchee committed:
cleanup

- prefix with `_` for internal functions
- fix `createPool` function issue
- add `_addPool` function

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
fixes

- assert global invariants in helpers
- add otherAccounts/_listAccounts pattern to

0xdavinchee committed:
fix some of the gda invariance issues

0xdavinchee committed:
testing cleanup

d10r committed:
merge dev

0xdavinchee committed:
add a few more actions

0xdavinchee committed:
fix tests

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Update getAbi.js

0xdavinchee committed:
fix build

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Update bundled-abi-contracts-list.json

d10r committed:
GDA tooling (#1602)

* gov scripts for governance update and transfer: make usable for prod governance

* simplify gov init

* ocd treatment: more consistent order of cof vs cif

* nix setup hint

* store version string in Resolver, fix contract change recognition for some factory related contracts

* fix code changed detection for contracts the SuperTokenFactory depends on

* new loader contracts on gda testnets

* rev string shall always have 8 chars

* store 16 chars of git revision

* [SUPPLYCHAIN] Update lerna again after nx@16.7.3 (#1589)

* [TOOLING] ops-scripts improvements for 1.7+ (#1586)

* gov scripts for governance update and transfer: make usable for prod governance

* simplify gov init

* ocd treatment: more consistent order of cof vs cif

* nix setup hint

* store version string in Resolver, fix contract change recognition for some factory related contracts

* rev string shall always have 8 chars

* store 16 chars of git revision

* Bump webpack in /sdk-redux-examples/sdk-redux-react-typecript (#1614)

Bumps [webpack](https://github.com/webpack/webpack) from 5.75.0 to 5.88.2.
- [Release notes](https://github.com/webpack/webpack/releases)
- [Commits](https://github.com/webpack/webpack/compare/v5.75.0...v5.88.2)

---
updated-dependencies:
- dependency-name: webpack
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Miao ZhiCheng <miao@superfluid.finance>

* [METADATA] 1594 | Add Subgraph endpoints to metadata (#1611)

* add to metadata

* update changelog

---------

Co-authored-by: Miao ZhiCheng <miao@superfluid.finance>

* Bump web3 from 1.10.0 to 1.10.1 (#1619)

Bumps [web3](https://github.com/ChainSafe/web3.js) from 1.10.0 to 1.10.1.
- [Release notes](https://github.com/ChainSafe/web3.js/releases)
- [Changelog](https://github.com/web3/web3.js/blob/v1.10.1/CHANGELOG.md)
- [Commits](https://github.com/ChainSafe/web3.js/compare/v1.10.0...v1.10.1)

---
updated-dependencies:
- dependency-name: web3
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Bump http-cache-semantics (#1620)

Bumps [http-cache-semantics](https://github.com/kornelski/http-cache-semantics) from 4.1.0 to 4.1.1.
- [Commits](https://github.com/kornelski/http-cache-semantics/compare/v4.1.0...v4.1.1)

---
updated-dependencies:
- dependency-name: http-cache-semantics
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Miao ZhiCheng <miao@superfluid.finance>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: 0xdavinchee <0xdavinchee@gmail.com>

0xdavinchee committed:
merge gda toolling

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
gdav1 forwarder tested

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
fix tests

0xdavinchee committed:
add required ISuperToken ABIs

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
include resolver/ISuperfluid abis

0xdavinchee committed:
fix build

0xdavinchee committed:
subgraph tests wip

pool
- handleFlowDistributionUpdated
- handleInstantDistributionUpdated
- handleDistributionClaimed
- handleMemberUnitsUpdated
- handlePoolConnectionUpdated
poolMember
- handleDistributionClaimed
- handleMemberUnitsUpdated

0xdavinchee committed:
pool distributor tests

0xdavinchee committed:
more echidna actions

0xdavinchee committed:
clean up hot fuzz + gda

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
add token actions

add some more token actions

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
add TOGA as account

0xdavinchee committed:
fix test cases

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
fix formatting

d10r commented on packages/ethereum-contracts/tasks/bundled-abi-contracts-list.json:
```diff
@@ -13,8 +13,8 @@
     "IConstantOutflowNFT", "ConstantOutflowNFT",
-    "IPoolAdminNFT",
-    "IPoolMemberNFT",
+    "IPoolAdminNFT", "PoolAdminNFT",
+    "IPoolMemberNFT", "PoolMemberNFT",
```
please also add `IGeneralDistributionAgreementV1` and `GeneralDistributionAgreementV1`

0xdavinchee replied:
done

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
fix formatting

0xdavinchee committed:
fix lint errors

0xdavinchee committed:
add callAgreement wrappers

0xdavinchee committed:
with ctx functions added

0xdavinchee committed:
cleanup

- reduce code size
- create SolvencyHelperLibrary
- finished supertokenv1library callback tests

0xdavinchee committed:
add to event

0xdavinchee committed:
fix gda address for avalanche-fuji

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
subgraph and sdk-core changes

0xdavinchee committed:
wip

- reduce code size
- cleaned up deployers
- schema additions
- subgraph add userData
- more cleanup incoming

d10r committed:
deploy script for gda forwarder & addition to framework deploy script

d10r committed:
added gdaV1Forwarder addresses for testnets

0xdavinchee committed:
fix subgraph tests

0xdavinchee committed:
Merge branch 'gda' of github.com:superfluid-finance/protocol-monorepo into gda

0xdavinchee committed:
deployment env

0xdavinchee committed:
change file structure

- move gda related contracts to `agreements/gdav1` folder
- same thing with the interface files, to `interfaces/agreements/gdav1`

0xdavinchee committed:
unnecessary suffering

- do not use the harness pattern in the future, it was completely unnecessary => using .prop pattern suffices
- code size optimization was unnecessary

0xdavinchee committed:
cleanup

- remove DelegatableTokenMonad
- fix up test

0xdavinchee committed:
fix hot fuzz breaking

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Update bundled-abi-contracts-list.json

0xdavinchee committed:
Merge branch 'gda' of github.com:superfluid-finance/protocol-monorepo into gda

hellwolf committed:
remove GDAv1Internal

0xdavinchee committed:
Delete GDAv1Internal.sol

0xdavinchee committed:
Merge branch 'gda' of github.com:superfluid-finance/protocol-monorepo into gda

0xdavinchee committed:
more memory for the job

0xdavinchee committed:
rename test functions

hellwolf reviewed (changes requested):
Pending audit and its fixes.

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
explicitly set parallel to false

0xdavinchee committed:
clean up build warnings

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
pass via env:

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
fix build

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
fixes attempt

- move event entity creation to the end of handler
- purposefully brick handleMemberUnitsUpdated to see if this is ever being hit

0xdavinchee committed:
fix build

0xdavinchee committed:
fix contract size issue

0xdavinchee committed:
use correct sig + add gda address to config

0xdavinchee committed:
gda address can be null

0xdavinchee committed:
fix broken unit tests

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
fix flakey tests

0xdavinchee committed:
Merge branch 'dev' into gda

d10r committed:
more consistent naming of verification related vars

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
pool config

- create pool with a poolConfig struct => this struct determines if unit transferability and distribution from any address is allowed
- tests added to ensure the new configs actually block the actions
- existing tests ensures that the configs don't break existing functionality
- fix flakey test in FoundrySuperfluidTester in _addAccount
- move SuperfluidPoolDeployerLibrary to gdaV1 folder

0xdavinchee committed:
cleanup

- comment cleanup
- code cleanup

0xdavinchee committed:
move pool config out of interface

0xdavinchee committed:
move struct to interface

0xdavinchee committed:
fix build

0xdavinchee committed:
fix build attempt

0xdavinchee committed:
fix build

hellwolf committed:
Merge branch 'dev' into gda

0xdavinchee committed:
Merge branch 'dev' into gda

hellwolf commented:
https://github.com/superfluid-finance/protocol-monorepo/pull/1357/commits/4ece1a3f4aff8b5a9cbf37118d261023960c0f0f was the first commit sent for audit.

0xdavinchee committed:
[ETHEREUM-CONTRACTS] [GDA] Fixes for GDA  (#1729)

* [ETHEREUM-CONTRACTS] BatchLiquidator: don't revert for non-transferrable SuperTokens (#1707)

* don't revert for non-transferrable SuperTokens

* add test for custom tokens revert on transfer

---------

Co-authored-by: Axe <seamac@gmail.com>

* patch getUnderlyingToken (#1718)

* Bump undici from 5.21.0 to 5.26.3 (#1719)

Bumps [undici](https://github.com/nodejs/undici) from 5.21.0 to 5.26.3.
- [Release notes](https://github.com/nodejs/undici/releases)
- [Commits](https://github.com/nodejs/undici/compare/v5.21.0...v5.26.3)

---
updated-dependencies:
- dependency-name: undici
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* add error hashes, use currentContext.timestamp

* use getHost.getTimestamp()

* Bump @babel/traverse from 7.21.3 to 7.23.2 (#1723)

Bumps [@babel/traverse](https://github.com/babel/babel/tree/HEAD/packages/babel-traverse) from 7.21.3 to 7.23.2.
- [Release notes](https://github.com/babel/babel/releases)
- [Changelog](https://github.com/babel/babel/blob/main/CHANGELOG.md)
- [Commits](https://github.com/babel/babel/commits/v7.23.2/packages/babel-traverse)

---
updated-dependencies:
- dependency-name: "@babel/traverse"
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* [SDK-CORE/ METADATA] SDK-Core No Governance Fix + Metadata Types (#1728)

* cleanup

* fix types

* remove tests flakiness

* remove forge.sh

* [SDK-CORE] GoodDollar sdk core fix (#1734)

* fix supertoken initialization for gooddollar

* bump version + update changelog

* use governance address from networkData

* gooddollar symbol

---------

Co-authored-by: Kaspar Kallas <kaspar@superfluid.finance>

* [ETHEREUM-CONTRACTS] make deploy script compatible with ethers v6 (#1730)

* make deploy script compatible with ethers v6

* more meaningful jsdoc

* add new functions for dev-scripts deployment

* update deploy.sh (#1738)

* Bump browserify-sign in /packages/sdk-core/previous-versions-testing (#1740)

Bumps [browserify-sign](https://github.com/crypto-browserify/browserify-sign) from 4.2.1 to 4.2.2.
- [Changelog](https://github.com/browserify/browserify-sign/blob/main/CHANGELOG.md)
- [Commits](https://github.com/crypto-browserify/browserify-sign/compare/v4.2.1...v4.2.2)

---
updated-dependencies:
- dependency-name: browserify-sign
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* Bump browserify-sign from 4.2.1 to 4.2.2 (#1739)

Bumps [browserify-sign](https://github.com/crypto-browserify/browserify-sign) from 4.2.1 to 4.2.2.
- [Changelog](https://github.com/browserify/browserify-sign/blob/main/CHANGELOG.md)
- [Commits](https://github.com/crypto-browserify/browserify-sign/compare/v4.2.1...v4.2.2)

---
updated-dependencies:
- dependency-name: browserify-sign
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* fix verification script

* change GDA forwarder address (needed to redeploy)

* change GDA forwarder address for polygon-mumbai

* fix tests to fuzz with different pool configs and fix distributeFlow blocked liquidation issue

* fix TOB-SUPERFLUID-2: Incorrect event emission in connectPool

* fix TOB-SUPERFLUID-5: Large encoded buffer amount could manipulate preceding field

* fix TOB-SUPERFLUID-6: Off-by-one gas left check

* fix broken test

* wrangle with reducing the code size of GDA...

* cleanup

* fix build + tests

* cleanup

* add update beacon proxy update paths

* fix broken deploy script, add tests, add transfer ownership of beacon in deploy script

* fix broken deploy script attempt 2

* [WORKFLOWS] Use nix in `handler.run-ethereum-contracts-script.yml` (#1745)

* handler.run-ethereum-contracts-script.yml

* add nix to handler.deploy-to-testnets.yml

* [ETHEREUM-CONTRACTS] App credit test (#1743)

* add app credit sanity test

* cleanup console.logs

* add cliName (#1748)

* doConnect != isConnected fixed

* remove extra whitespace

* add assertEq in SFGovII test and import PoolConfig in ISuperfluid for easy access

* fix build

* fuzzing fix

* EXPECT BREAKAGE IN FUZZ

* fix build but echidna should break

* undo breakage

* fix the test

* bump sdk-core version, fix sdk-core operation functions, fix subgraph mapping

* fix unit tests

* fix again

* hot fuzz additions

* [ETHEREUM-CONTRACTS] Fix canary build (#1742)

* allow listing non-standard SuperTokens

* fix canary build, fixes #1633

* distributeFlow: fix order of args to be consistent

* fix build

* add fix back in

* [ETHEREUM-CONTRACTS] new resolver and loader, updated and bumped metadata, refs #1004 (#1750)

* new resolver and loader, updated and bumped metadata, refs #1004

* updated changelog

* gda version of loader

* fix build

* new resolver and loader, refs #1004

* remove getIsListed workaround assuming resolver exists (#1751)

* fix broken unit test

* fix gda logic

* [ETHEREUM-CONTRACTS] new resolver & loader address for mainnets (#1752)

* new resolver & loader address for mainnets

* bumped metadata version

* remove unimplemented function from yaml (#1753)

* map the name from subgraph to unknown entity (#1754)

* fix mapping

* add total units

* missing import

* missing import pt 2

* [WORKFLOW] Subgraph deloy all networks (#1760)

* added VENDOR_NETWORKS variable

* decaled variable

* fix mapping (#1758)

* added new supported subgraphs (#1761)

* remove duplicate verification

* subgraph mapping addition

* type fix (#1771)

* readme fix

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Didi <git@d10r.net>
Co-authored-by: Axe <seamac@gmail.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Kaspar Kallas <kaspar@superfluid.finance>
Co-authored-by: Miao ZhiCheng <miao@superfluid.finance>
Co-authored-by: Momodu Afegbua <afegbuaismaila@gmail.com>

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
bump versions

0xdavinchee committed:
Merge branch 'dev' into gda

0xdavinchee committed:
merge fix

0xdavinchee committed:
upload audit report

0xdavinchee commented:
Use `git diff 4ece1a3f4aff8b5a9cbf37118d261023960c0f0f.. packages/ethereum-contracts/contracts` to see the changes in the contract code since the audit commit hash.

## High Level Summary of Changes

### GeneralDistributionAgreementV1
- The representation of totalBuffer is modified to ensure proper data fitting in a 256-bit field.
- `realtimeBalanceVectorAt` removed
- `PoolConnectionUpdated` event only emitted if the connection was changed

### SuperfluidPool
The method for obtaining timestamps and checking member connections is updated to use Superfluid framework methods instead of Ethereum's native functionalities.

0xdavinchee commented:
one more comment for good measure :)

0xdavinchee committed:
changes since audit file added

hellwolf commented on packages/ethereum-contracts/audits/changes-since-ToB-2023-audit.md:
```diff
@@ -0,0 +1,13 @@
+# Changes Since ToB 2023 Audit
+
+Use `git diff 4ece1a3f4aff8b5a9cbf37118d261023960c0f0f.. packages/ethereum-contracts/contracts` to see the changes in the contract code since the audit commit hash.
```
```
       git log [<options>] [<revision-range>] [[--] <path>...]
```

could work too

hellwolf replied:
`git diff|log audit-tob-2023..dev`

hellwolf reviewed (approved):
OMG, it's happening! But there is no way to comment on this without making it anti-climatic!
